enhancements/cluster-logging/fluentd-tweaks.md

```diff
@@ -0,0 +1,166 @@
+---
+title: cluster-logging-log-forwarding
+authors:
+  - "@alanconway"
+reviewers:
+approvers:
+creation-date: 2020-05-27
+status: provisional
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# Fluentd Peformance Tuning
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+### Naming of "tweaks"
+
+I've called these settings 'tweaks' (like the Gnome Tweaks application) to
+identify them clearly as things that most users don't need to set.  If there are
+objections to that, possible alternate names: 'knobs', 'hints', 'advanced',
+'expert', 'tuning' ... other?
+
+### Where to put the tweaks
+
+Below I describe tweaks as part of the Collector API, acting as defaults for all outputs. This is the way we current use environment variables.
+However while writing this I came to the conclusion that they may fit better as part of the ClusterLogForwarder.Output.
+
+Pro:
+* Allows different settings per output: different outputs may have different QoS (remote vs. local, important to keep vs. may be dropped)
+* Follows our own proposal to provide generic retry options (not implemented for GA, was in the original forwarder design)
+* Follows fluentd model - `<buffer>` settings are part of the `<output>` section.
+
+Con:
+* Need to clearly separate *tweaks* (local sender behavior) from *destination* configuration (URL, protocol etc.)
+  E.g. a syslog output with tweaks could have a `syslog` *destination* section and `fluentd` tweaks section.
+
+## Summary
+
+Allow selected fields from the fluentd `<buffer>` configuration section to be
+specified for logging, for performance-tuning purposes. These settings are
+referred to as `tweaks` to clearly identify that they are for advanced
+performance-tuning use only.
+
+Tweaks are:
+* *not* relevant to most users, default settings should give good general performance.
+* only for *advanced* users with detailed knowledge of `fluentd` configuration and performance.
+* only for performance tuning, they have no effect on functional aspects of logging.
+
+Tweaks have weaker forward-compatibility guarantees than the rest of the API.
+In future releases:
+  - `fluentd` tweaks *may not apply* if the collector implementation changes.
+  - values set in tweaks *may be overridden* by settings in the collector or forwarder APIs.
+  - future collectors may or may not have a tweaks section
+  - future collectors may not have equivalent tweaks to `fluentd`.
+
+This will be presented in the `collectors.logging.openshift.io` API as a
```

Comment:
This API is currently only devpreview that is availble to provide alternate collectors without support.  Do we want to promote this or should the changes be introduced into the CL api?

Reply:
Need to settle the previous question of where this belongs first - if we do put in Forwarder there's still time before GA.

Reply:
Strong vote into the CL API for all the reasons pointed out. Let's not mix them into the Forwarder API

Reply:
Moving to ClusterLogging API solves both these problems.